### PR TITLE
create taproot descriptor template

### PR DESCRIPTION
**Pull request by vladimirfomene**
### Description

This PR solves bitcoindevkit/bdk_wallet#169. This PR adds a P2TR
descriptor template and a BIP86 taproot
descriptor template. With this, users
can now create a taproot descriptor with templates.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Review by w0xlt — COMMENTED** [review]
Concept ACK

File: src/descriptor/template.rs
Comment (rajarshimaitra) [inline src/descriptor/template.rs]: I was a bit confused with this at first. Why a `Mainnet` descriptor matches with a `Regtest` address? That's because the first `network` is used for setting the 2nd derivation index of `Bipxx`, and the second network is used for the address prefix.

As a result, if someone actually tries with the same `Xpriv` but builds with `build(Network::Regtest)` and derives the addresses, they will not match up with the test vector. 

So does it make sense to change the first network into `regtest` in the build call, or change the second network to `mainnet`? The first option is a much smaller changeset and can be done in a separate PR.

**Review by rajarshimaitra — APPROVED** [review]
Review ACK 242f5bd951eef624cc7def4b654b9190bb422b8f

Everything looks good to me. Just one observation.

**Comment by notmandatory** [comment]
@vladimirfomene please rebase to pickup the new MSRV change from bitcoindevkit/bdk#842.

File: src/descriptor/template.rs
Comment (notmandatory) [inline src/descriptor/template.rs]: Should be updated in #838:
```suggestion
/// This assumes that the key used has already been derived with `m/86'/0'/0'` for Mainnet or `m/86'/1'/0'` for Testnet.
```

**Comment by notmandatory** [comment]
Would be good to get this one in before merging bdk_core work.

**Comment by vladimirfomene** [comment]
I will work on this tonight, I will be ready tomorrow

File: src/descriptor/template.rs
Comment (vladimirfomene) [reply src/descriptor/template.rs]: @rajarshimaitra , you are right this is a bit confusing. I will opt for your first option. I will option a small PR for it before end of the week. I will open an issue that helps me keep track of it.

File: src/descriptor/template.rs
Comment (vladimirfomene) [reply src/descriptor/template.rs]: Noted

**Comment by notmandatory** [comment]
Thanks for all your work on this but we've gotten to the point where I'd like to hold off on merging any new features and only merge critical bug fixes to the `release/0.27` branch. 

Can you rebase this after the new `bdk_core_staging` stuff in bitcoindevkit/bdk#793 is merged to `master` branch? then we can merge it in for one of the upcoming 1.0.0-alpha releases.

**Comment by danielabrozzoni** [comment]
Hey, can you please rebase this one on master?

**Comment by notmandatory** [comment]
I see we didn't get merged in before the big bdk 1.0 switch. It It looks like it's ready to go, can you rebase this or update it for the new `master` branch? 

I was looking for this because @reez and I want to use a TR template for the iOS example app we're doing.

**Comment by notmandatory** [comment]
After this is merged to master I'll get it back-ported to a maintenance release.

**Comment by vladimirfomene** [comment]
@notmandatory , this is good to go. I have created an issue bitcoindevkit/bdk_wallet#149 for fixing the issue raised by @rajarshimaitra

**Review by notmandatory — APPROVED** [review]
ACK e30919ba3a7941e6f54e2f95e94533c05415d831

**Pull request merged**
